trl/trainer/ppo_trainer.py

```diff
@@ -305,21 +309,19 @@ def __init__(
             self.lr_scheduler,
         )
         if is_deepspeed_used:
-            # 8 bit models are already set on the correct device
+            # Quantized models are already set on the correct device
             if not self.is_peft_model and not (
                 getattr(self.ref_model.pretrained_model, "is_loaded_in_8bit", False)
                 or getattr(self.ref_model.pretrained_model, "is_loaded_in_4bit", False)
             ):
-                # DS integration only allows for single model and as `ref_model` is only used for
-                # `KL divergence loss`,i.e, in eval model, just have it be on the respective device and
-                # there is no need to pass it to the `accelerator.prepare` call
-                self.ref_model = self.ref_model.to(self.accelerator.device)
-
-            # this hack seems to be needed for DS stage 3 to work
-            if self.accelerator.state.deepspeed_plugin.zero_stage == 3:
-                self.model.train()
+                # If ZeRO-3 is used, we shard both the active and reference model.
+                # Otherwise, we assume the reference model fits in memory and is copied to each device.
+                if self.accelerator.state.deepspeed_plugin.zero_stage == 3:
+                    self.ref_model = self._prepare_deepspeed_zero3(self.ref_model)
+                else:
+                    self.ref_model = self.ref_model.to(self.accelerator.device)
         else:
-            self.ref_model = self.accelerator.prepare(self.ref_model)
+            self.ref_model = self.accelerator.prepare_model(self.ref_model, evaluation_mode=True)
```

Comment:
Since the reference model is only for inference, let's make sure we put it in eval mode